### ISSUE TITLE
Failure to recognise function type for static member function template in class template

### ISSUE THREAD

**Issue by cuavas** [llvm/llvm-project#91362]
We noticed this as code using sol2 failing to compile with clang 18.  It was reported to that project here: ThePhD/sol2#1581 (also as duplicate ThePhD/sol2#1588).

It’s demonstrated by this reduced test case:
```c++
#include <type_traits>

template <typename T>
struct class_tmpl {
    template <bool B> static int call_e() { return 0; }
    template <bool B> static int call_ne() noexcept { return 0; }
    template <bool B> static int call() noexcept(std::is_nothrow_copy_assignable<T>::value) { return 0; }
};

int main(int argc, char *argv[])
{
    using function_ptr = int (*)();
    function_ptr f1 = &class_tmpl<int>::call_e<false>;
    function_ptr f2 = &class_tmpl<int>::call_ne<false>;
    function_ptr f3 = &class_tmpl<int>::call<false>;
    return 0;
}
```

With clang 18 (including clang 18.1), compilation fails with `-std=c++14` or `-std=c++17` with the following error:
```
test.cpp:15:24: error: address of overloaded function 'call' does not match required type 'int ()'
   15 |     function_ptr f3 = &class_tmpl<int>::call<false>;
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
test.cpp:7:34: note: candidate template ignored: substitution failure [with B = false]
    7 |     template <bool B> static int call() noexcept(std::is_nothrow_copy_assignable<T>::value) { return 0; }
      |                                  ^
1 error generated.
```

When the `noexcept` specification uses an expression that depends on the class template arguments, clang reports a substitution failure.  The code compiles with clang 17 and earlier, as far back as clang 6 at least.

**Comment by cuavas** [comment]
Even simpler test case (without using the `<type_traits>` header) that fails as C++17:
```c++
template <typename T>
struct class_tmpl {
    template <bool B> static void call_e() { }
    template <bool B> static void call_ne() noexcept { }
    template <bool B> static void call() noexcept(sizeof(T) > 1) { }
};

int main(int argc, char *argv[])
{
    using function_ptr = void (*)();
    function_ptr f1 = &class_tmpl<int>::call_e<false>;
    function_ptr f2 = &class_tmpl<int>::call_ne<false>;
    function_ptr f3 = &class_tmpl<int>::call<false>;
    return 0;
}
```

Error with clang 18.1.4:
```
% clang++ -std=c++17 test.cpp
test.cpp:13:24: error: address of overloaded function 'call' does not match required type 'void ()'
   13 |     function_ptr f3 = &class_tmpl<int>::call<false>;
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
test.cpp:5:35: note: candidate template ignored: substitution failure [with B = false]
    5 |     template <bool B> static void call() noexcept(sizeof(T) > 1) { }
      |                                   ^
1 error generated.
```

**Comment by zyn0217** [comment]
This should have been resolved by #90760.
(I'll close the issue once that gets applied to godbolt. ATM we encounter a crash with an assertion build for the case above: https://gcc.godbolt.org/z/vcEr5sroY)

**Comment by cuavas** [comment]
It seems godbolt has updated to a build that successfully compiles the code.

**Comment by zyn0217** [comment]
Yeah, so let's close the issue and feel free to re-open it if you find other cases.
